src/ui.h

```diff
@@ -536,22 +536,38 @@ class pointmenu_cb : public uilist_callback
 void kill_advanced_inv();
 
 /**
- * Helper for typical arrow key UI list navigation with wrap-around
- * Add 1/-1 to val, then wrap to the range [0,size)
+ * Helper for typical UI list navigation with wrap-around
+ * Add delta to val, then wrap to the range [0,size)
  */
 template<typename V, typename S>
-inline typename std::enable_if < !std::is_enum<V>::value, V >::type increment_and_wrap( V val,
-        bool inc, S size )
+inline typename std::enable_if < !std::is_enum<V>::value, V >::type
+increment_and_wrap( V val, int delta, S size )
 {
     if( size == 0 ) {
         return 0;
     }
-    if constexpr( std::is_unsigned_v<V> ) {
-        if( !inc && val == 0 ) {
-            return size ? size - 1 : 0;
-        }
+    // wrap to last
+    if( val <= 0 && delta < 0 ) {
+        return size - 1;
+    }
+    // wrap to first
+    // Templating of existing `unsigned int` triggers linter rules against `unsigned long`
+    // NOLINTNEXTLINE(cata-no-long)
+    if( val >= static_cast<V>( size ) - 1 && delta > 0 ) {
+        return 0;
     }
-    return modulo( val + ( inc ? 1 : -1 ), size );
+    return std::clamp<V>( val + delta, 0, size - 1 );
```

Comment:
So this will clamp to 0, size, then only wrap on the next increase? That's pretty counterintuitive.

Reply:
If that is the intended behavior,  it needs a brief comment clarifying it,  and/or the function needs a new name. clamp_then_wrap() gets the idea across but it's awkward as hell.

Reply:
@kevingranade is it good enough now?

Reply:
I also thought about splitting them:
`inc_clamp_wrap( V val, int delta, S size )` needs the clamp in name, as it clamps
`inc_clamp_wrap( V val, bool inc, S size )` and `inc_clamp_wrap( T val, I inc, T size )` don't need clamp in name, as they only increase or decrease by one, so they can never be clamped. So rename them to `inc_wrap`.
Then, there would be three names: `inc_clamp_wrap`, `inc_clamp`, `inc_wrap`

Or maybe it would be better to make them all into one function? with parameters `bool clamp` `bool wrap`?
in there break into four cases
```
!clamp && !wrap {static assert("has to choose at least one `clamp` or `wrap`")}
clamp && !wrap { the inc_clamp code }
!clamp && wrap { the inc_clamp_wrap code - BUT not stopping at bounds }
clamp && wrap { the inc_clamp_wrap code }
```

Reply:
I consider this solved.